src/frontends/pytorch/src/op/unfold.cpp

```diff
@@ -0,0 +1,97 @@
+    auto ndim = context.mark_node(std::make_shared<opset8::Reshape>(ndim_tmp, const_1, false));
+    auto dimension_scalar = context.mark_node(std::make_shared<opset8::Reshape>(dimension, const_1, false));
+    auto dimension_plus_1_scalar =
+        context.mark_node(std::make_shared<opset8::Reshape>(dimension_plus_1, const_1, false));
+    auto perm_begin =
+        context.mark_node(std::make_shared<opset8::Range>(const_0, dimension_scalar, const_1, element::i64));
+    auto perm_end =
+        context.mark_node(std::make_shared<opset8::Range>(dimension_plus_1_scalar, ndim, const_1, element::i64));
+    auto perm = context.mark_node(std::make_shared<opset8::Concat>(OutputVector{perm_begin, perm_end, dimension}, 0));
+
+    // body parameters
+    auto input_param = std::make_shared<opset8::Parameter>(element::f32, PartialShape::dynamic());
+    auto low_ind_param = std::make_shared<opset8::Parameter>(element::i64, PartialShape::dynamic());
+    auto hi_ind_param = std::make_shared<opset8::Parameter>(element::i64, PartialShape::dynamic());
+    auto perm_param = std::make_shared<opset8::Parameter>(element::i64, PartialShape::dynamic());
+    auto iter_param = std::make_shared<opset8::Parameter>(element::i64, PartialShape::dynamic());
+
+    // body
+    auto iter_plus_1 = context.mark_node(std::make_shared<opset8::Add>(iter_param, const_1_list));
+    auto low_ind_curr_iter = context.mark_node(
+        std::make_shared<opset8::Slice>(low_ind_param, iter_param, iter_plus_1, const_1_list, const_0_list));
+    auto hi_ind_curr_iter = context.mark_node(
+        std::make_shared<opset8::Slice>(hi_ind_param, iter_param, iter_plus_1, const_1_list, const_0_list));
+    auto slice = context.mark_node(
+        std::make_shared<opset8::Slice>(input_param, low_ind_curr_iter, hi_ind_curr_iter, const_1_list, dimension));
+    auto transpose = context.mark_node(std::make_shared<opset8::Transpose>(slice, perm_param));
+    auto unsqueeze = context.mark_node(std::make_shared<opset8::Unsqueeze>(transpose, dimension));
+    auto body =
+        std::make_shared<Model>(OutputVector{unsqueeze},
+                                ParameterVector{iter_param, input_param, low_ind_param, hi_ind_param, perm_param});
+
+    // number of iterations
+    auto low_indices_count = context.mark_node(std::make_shared<opset8::ShapeOf>(low_indices));
+    auto hi_indices_count = context.mark_node(std::make_shared<opset8::ShapeOf>(hi_indices));
+    auto iterations_count = context.mark_node(std::make_shared<opset8::Minimum>(low_indices_count, hi_indices_count));
+    auto iterations_count_scalar =
+        context.mark_node(std::make_shared<opset8::Reshape>(iterations_count, const_1, false));
+    auto iter_values =
+        context.mark_node(std::make_shared<opset8::Range>(const_0, iterations_count_scalar, const_1, element::i64));
+    auto tensor_iterator = std::make_shared<opset8::TensorIterator>();
```

Comment:
If we try to express the code with Gather and some arithmetic on indices without any loop involvement, will it be much more complicated? The fact is that GPU won't quickly enable such operation decomposition via the loop if dynamic shapes are involved. But if it would be expressed through Gather it will be supported more quckly. Unfortunately we are bit behind in support of control flow ops for new scenarios (like dynamic shapes).

Reply:
I find it rather hard to decompose this operator without usage of a Loop/TensorIterator operator. Of course we could use Gather and some arithmetic on indices but it will require us to use Loop/TensorIterator in the other place. The problem is that slices of data used in each step of loop are generated in a complicated way.

Reply:
After further investigation I found out, that it is possible to implement what you have proposed, so I will work on new implementation (without usage of TensorIterator).